common/environment/environment.cpp

```diff
@@ -1336,12 +1536,221 @@ void CLocalEnvironment::clearCache()
         p.setown(conn->getRoot());
     }
     cache.kill();
-    machineCacheBuilt = false;
-    dropZoneCacheBuilt = false;
+    numOfDropzonesByComputer.clear();
+    init();
     resetPasswordsFromSDS();
 }
 
+IConstDropZoneInfo * CLocalEnvironment::findDropZoneByAddress(const char * netaddress, const char *targetFilePath) const
+{
+    IConstDropZoneInfo * retVal = NULL;
+
+    Owned<IConstMachineInfo> machineInfo = getMachineByAddress(netaddress);
+    if (!machineInfo)
+        return retVal;
+
+    SCMStringBuffer machineName;
+    machineInfo->getName(machineName);
+    const char * pmachineName = machineName.str();
+    Owned<IConstDropZoneInfoIterator> zoneIt= getDropZoneIteratorByComputer(pmachineName);
+    LOG(MCdebugInfo, unknownJob, "Check remote drop zones (%d) by iterator on '%s':", zoneIt->count(), pmachineName);
+
+    SCMStringBuffer dropzonePath;
+    const char * pdropzonePath = NULL;
+    unsigned dropzonePathLen = _MAX_PATH + 1;
+
+    ForEach(*zoneIt)
+    {
+        SCMStringBuffer dropZoneName;
+        zoneIt->query().getName(dropZoneName);
+
+        SCMStringBuffer dropZoneDir;
+        zoneIt->query().getDirectory(dropZoneDir);
+        const char * pdropzoneDir = dropZoneDir.str();
+        LOG(MCdebugInfo, unknownJob, "Drop zone name:'%s', directory:'%s'", dropZoneName.str(), pdropzoneDir);
+
+        // Check target file path starts with this Drop zone path
+        // the drop zone paths can be nested (nothing forbids it) like
+        // dz1: /a/b/c/d
+        // dz2: /a/b/c
+        // dz3: /a/b
+        // check all and select the shortest
+
+        if ( strncmp(pdropzoneDir, targetFilePath, strlen(pdropzoneDir)) == 0)
```

Comment:
I think it can't crash. If the targetFilePath is sorter, than pdropzoneDir, then the comparison stopped at the shorter string terminator zero byte (based on it is not equal with the char at the same position in other string).